Updating build system for use in CommonJS

6utt3rfly opened the pull request:
Trying to use jsep in commonjs wasn't working because it was being treated as a module.
- Separate dist folders for each build. Add overriding package.json file in cjs folder to set `type: 'commonjS'`. (Note: there are difficulties with trying to do this using `cat` in an npm script with {} and some Operating Systems. Simple solution is to copy a file.
- Updated ReadMe
- Added tests to lint script and fixed lint errors

6utt3rfly committed:
Updating build system for use in CommonJS
- Separate dist folders for each build. Add overriding package.json file in cjs folder to set `type: 'commonjS'`. (Note: there are difficulties with trying to do this using `cat` in an npm script with {} and some Operating Systems. Simple solution is to copy a file.
- Updated ReadMe
- Added tests to lint script and fixed lint errors

6utt3rfly commented on README.md:
```diff
@@ -23,35 +23,50 @@ The jsep built files will be in the build/ directory.
+const parse_tree = Jsep.parse('1 + 1');
+
+// or: (NOTE: versions prior to 1.x didn't require `.default`)
+const jsep = require('jsep').default;
```
@LeaVerou or @EricSmekens ... not sure if we want to leave it like this? Or commonjs has to require from a different folder to get the named export?

LeaVerou replied:
Yeah I think it might be better if we could avoid this, but no strong opinion.

6utt3rfly replied:
I was able to fix this ... it's a double build though to avoid combining named and default exports 🤷🏻‍♀️  (for this small a project, seems okay?)

LeaVerou commented on README.md:
```diff
@@ -23,35 +23,50 @@ The jsep built files will be in the build/ directory.
+<script type="module">
+import { Jsep } from '/PATH/TO/jsep.min.js';
+const parsed = Jsep.parse('1 + 1');
+</script>
```
I think here we might want to show something closer to the original example, using the IIFE bundle. 
Or maybe both.

6utt3rfly replied:
Sounds good - I think I have it building the IIFE so it's more compatible now

LeaVerou commented on README.md:
```diff
@@ -23,35 +23,50 @@ The jsep built files will be in the build/ directory.
-let parse_tree = jsep("1 + 1");
+// ESM:
+import { Jsep } from 'jsep';
+const parse_tree = Jsep.parse('1 + 1');
```
What is the advantage of this for our users over just using `jsep()`?

6utt3rfly replied:
I'll go back to both. I've been reading tonight a lot about the "evils" of default exports ... so I think our preferred usage should be to use the named export, but I've got it working for both CJS and ESM with both the default and named export now (without needing '.default')

LeaVerou replied:
I always try to keep an open mind, so feel free to send these "evils" my way, but so far I'm entirely unconvinced by the arguments against default exports. The ones I've seen seem to mainly center around inadequacies of tooling. IMHO it's the tooling that needs to be fixed, not default exports that should be avoided. There is a **very** clear usability benefit to default exports. Not only can the module be imported without looking up its documentation to remember what name to use, but also it offers a clear entry point, that *this* object is the main thing to use. And the syntax is simpler too. These may seem trivial for experienced developers, but they can make a world of difference to beginners who get easily discouraged and frustrated.

LeaVerou replied:
I was waiting for a response to this, not sure if it still applies?

6utt3rfly replied:
ah sorry! I thought it was resolved by providing both named and default exports?
From what I understand:
- using default imports prevents tree-shaking if there are parts of a file that are unused
- mixing named and default exports seems to be discouraged (which is why the build is more difficult for both cjs and esm). But I like that it allows both use cases.
- default imports can lead to inconsistent naming in a code base and may make large-scale refactoring more difficult.
- intellisense may be more problematic with default imports (tooling problem, I know)
- named exports are easier in export aggregation (i.e. barrel files)

In my work, I use a lot of both CJS default and named require statements, but mostly named imports in typescript. (Trying to get away from import * as _ from 'lodash'!). I've found that unit testing named exports can sometimes be easier. But overall, I don't have a strong opinion either way.

LeaVerou commented on README.md:
````diff
@@ -23,35 +23,50 @@ The jsep built files will be in the build/ directory.
 ```javascript
 // Add a custom ^ binary operator with precedence 10
-jsep.addBinaryOp("^", 10);
+Jsep.addBinaryOp("^", 10);
````
Maybe we want to include import statements here to make it clear what these methods are called on?

6utt3rfly replied:
I've reverted these changes, and I think we can remove the identifier chars in a separate PR if we still want? It's newer, but isn't pretty minimal anyway. Depends how much of a breaking change the next release should be?

LeaVerou commented on README.md:
````diff
@@ -60,10 +75,10 @@ You can add or remove additional valid identifier chars. ('_' and '$' are alread
 ```javascript
 // Add a custom @ identifier
-jsep.addIdentifierChar("@");
+Jsep.addIdentifierChar("@");
````
Didn't we resolve to remove these?

LeaVerou commented on rollup.config.js:
```diff
@@ -6,19 +6,17 @@ const { version } = require('./package.json');
-
 	return {
-		file: `dist/jsep${suffix}.js`,
+		file: `dist/${format}/jsep.${type}.js`,
```
Please note that `suffix` and `type` are not the same thing. I think there should be a "canonical" `jsep.js` with no format specifier, and I think that should be the ESM version. 
I also think it's a little strange if `/dist/jsep.js` points to nothing because everything is in subfolders. What do you think about `/dist/` containing the ESM build and then all other formats (CJS, IIFE etc) to be in their own subfolders? That would still maintain the same folder structure you have here for them, while also providing nicer URLs for the ESM bundle(s).

6utt3rfly replied:
Fair point. I was trying to get both CJS and ESM working properly without having to specify the full path and it took a few tries to get it working. Since it only requires the one overriding package.json I've put it back to the original dist file structure

LeaVerou commented on test/tests.js:
```diff
@@ -1,6 +1,7 @@
 (function() {
+/* eslint-disable brace-style */
 var binops = {
 	"+" : function(a, b) { return a + b; },
```
Unpopular opinion: If your linting rules force you to use comments to avoid warnings, then either the code or the linting rules need to be revised. 
Thankfully, I think in this case we can just refactor the code to use arrow functions, which would be more readable too, this kind of thing is exactly why they exist.

LeaVerou reviewed (changes requested):
Mostly LGTM, see comments!

6utt3rfly commented on package.json:
```diff
@@ -33,11 +41,11 @@
-		"default": "npm run build & npm run test & npm run docco",
-		"build": "npx rollup -c",
+		"default": "npm run lint && npm run build && npm run test && npm run docco",
+		"build": "npx rollup -c && cp package-cjs.json dist/cjs/package.json",
```
there's a copy plugin for rollup if that's preferred?

LeaVerou replied:
Why are we copying this? Why not just have it in `/dist/cjs/` in the first place?

6utt3rfly replied:
the `dist` folder isn't committed (and is typically deleted and rebuilt anyway). I tried to use `cat > ...` but it didn't like the `{}` (OS-specific I think)

LeaVerou replied:
Fair enough. I guess it would be a hassle to restructure our `.gitignore` so that that particular file is preserved.

6utt3rfly committed:
Update build system
- default export now includes static fields. Will allow for removing modifier methods if we like
- default export now adds Jsep as well to support `const { Jsep } = require('jsep')`
- removed eslint disable (switched to arrow functions)
- restored dist file folder structure, and some readme changes
- rollup now does 2 builds so it can use named + default export for ESM, and just the default export for IIFE and CJS (without having to use a .default property)

LeaVerou commented on README.md:
````diff
@@ -23,9 +23,9 @@ The jsep built files will be in the build/ directory.
-<script src="/PATH/TO/jsep.min.js"></script>
-...
+```html
+<script src="/PATH/TO/jsep.iife.min.js"></script>
````
I do think we should mention ESM too. They're not only useful for Node. Not sure if they should be before or after IIFE, I'll leave it up to you. I could argue this both ways: ESM should be first because it's the "canonical" version, or IIFE should be first because ESM might scare novices away.
Also, if we're including `<script>` elements, we should probably include one around the sample code.

LeaVerou commented on rollup.config.js:
```diff
@@ -6,46 +6,63 @@ const { version } = require('./package.json');
 
 	return {
-		file: `dist/jsep${suffix}.js`,
+		file: `dist/${folder}/jsep${suffix}.js`,
```
Note that if `folder` is empty, this would include a double slash (`//`). Which I guess works fine, but just wanted to point it out.

6utt3rfly replied:
I was thinking of that when I wrote that, but then forgot to update it! Fixed now :-)

6utt3rfly committed:
Fix dist//
update readme example for <script> and ESM

6utt3rfly commented:
Not sure if there's still a change needed @LeaVerou ? Or just waiting for a second review by @EricSmekens ?

LeaVerou commented:
Sorry I dropped the ball, thanks for the ping! I approved it now, so you could merge it yourself whenever you want (I'm told this is the new pattern so I figured I'd try it)
